lib/zipkin-tracer/rack/zipkin-tracer.rb

```diff
@@ -25,59 +27,33 @@ class RackHandler
     def initialize(app, config = nil)
       @app = app
       @lock = Mutex.new
-
-      config = Config.new(app, config)
-      SuckerPunch.logger = config.logger
-      adapter = config.adapter
-
-      ::Trace.tracer = case adapter
-        when :json
-          require 'zipkin-tracer/zipkin_json_tracer'
-          ::Trace::ZipkinJsonTracer.new(json_api_host: config.json_api_host, traces_buffer: config.traces_buffer)
-        when :scribe
-          require 'zipkin-tracer/careless_scribe'
-          ::Trace::ZipkinTracer.new(CarelessScribe.new(config.scribe_server), config.scribe_max_buffer)
-        when :kafka
-          require 'zipkin-tracer/zipkin_kafka_tracer'
-          ::Trace::ZipkinKafkaTracer.new(zookeepers: config.zookeeper)
-        else
-          ::Trace::NullTracer.new
-      end
-
-      ip_format = config.adapter == :json ? :string : :i32
-      ::Trace.default_endpoint = ::Trace::Endpoint.make_endpoint(
-        nil, # auto detect hostname
-        config.service_port,
-        service_name(config.service_name),
-        ip_format
-      )
-      ::Trace.sample_rate = config.sample_rate
-
-      @config = config
+      @config = Config.new(app, config).freeze
+      @tracer = TracerFactory.new.tracer(@config)
     end
 
     def call(env)
-      # skip certain requests
-      return @app.call(env) if filtered?(env) || !routable_request?(env)
-
-      whitelisted = force_sample?(env)
-      id = get_or_create_trace_id(env, whitelisted) # note that this depends on the sample rate being set
-      tracing_filter(id, env, whitelisted) { @app.call(env) }
+      zipkin_env = ZipkinEnv.new(env, @config)
+      @trace_id = zipkin_env.trace_id
```

Comment:
mmm I see it's being implicitly used by `trace!` and `trace_request_information`, wondering if the trace_id var shouldn't be passed as an arg. your call.